File: updateLiveScores.js
#!/usr/bin/env node
/**
 * NHL Live Score Updater (Node.js version)
 * =========================================
 * Fetches live scores from NHL API and updates BOTH JSON file AND Firebase.
 * Does NOT rescrape odds - only updates scores 2-3 times during game time.
 * 
 * Usage:
 *   node updateLiveScores.js                    # Update today's games
 *   node updateLiveScores.js --date 2025-10-23  # Specific date
 *   node updateLiveScores.js --continuous       # Run every 5 minutes
 * 
 * No additional dependencies needed - uses built-in fetch & fs
 */

import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import admin from 'firebase-admin';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Initialize Firebase Admin SDK (simplified service account like generateExpertAnalysis.js)
const serviceAccount = {
  project_id: process.env.VITE_FIREBASE_PROJECT_ID,
  client_email: process.env.FIREBASE_CLIENT_EMAIL,
  private_key: process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, '\n')
};

// Initialize Firebase if credentials are available
if (!admin.apps.length && serviceAccount.project_id && serviceAccount.client_email && serviceAccount.private_key) {
  try {
    admin.initializeApp({
      credential: admin.credential.cert(serviceAccount)
    });
    console.log('✅ Firebase Admin initialized');
  } catch (error) {
    console.warn('⚠️  Could not initialize Firebase Admin:', error.message);
    console.warn('   Scores will only be saved to JSON file');
  }
} else if (!admin.apps.length) {
  console.warn('⚠️  Missing Firebase credentials in environment variables');
  console.warn('   Required: VITE_FIREBASE_PROJECT_ID, FIREBASE_CLIENT_EMAIL, FIREBASE_PRIVATE_KEY');
  console.warn('   Scores will only be saved to JSON file');
}

const db = admin.apps.length > 0 ? admin.firestore() : null;

class NHLScoreUpdater {
  constructor(outputDir = 'public') {
    this.outputDir = path.join(__dirname, outputDir);
    this.outputFile = path.join(this.outputDir, 'live_scores.json');
    this.nhlApiBase = 'https://api-web.nhle.com/v1';
    this.db = db;
  }

  /**
   * Fetch detailed live game data including clock
   */
  async fetchLiveGameDetails(gameData) {
    try {
      const url = `${this.nhlApiBase}/gamecenter/${gameData.gameId}/play-by-play`;
      const response = await fetch(url);
      
      if (!response.ok) {
        console.warn(`⚠️  Could not fetch live details for game ${gameData.gameId}`);
        return;
      }
      
      const data = await response.json();
      
      // Update period and clock from detailed data
      if (data.periodDescriptor) {
        gameData.period = data.periodDescriptor.number || gameData.period;
        gameData.periodDescriptor = data.periodDescriptor.periodType || gameData.periodDescriptor;
      }
      
      if (data.clock) {
        gameData.clock = data.clock.timeRemaining || '';
      }
    } catch (error) {
      console.warn(`⚠️  Error fetching live game details: ${error.message}`);
    }
  }

  /**
   * Fetch game data from NHL API for specified date
   */
  async fetchGames(date = null) {
    // CRITICAL FIX: Use ET date if no date provided
    if (!date) {
      const etDateStr = new Date().toLocaleString('en-US', {
        timeZone: 'America/New_York',
        year: 'numeric',
        month: '2-digit',
        day: '2-digit'
      });
      const [month, day, year] = etDateStr.split('/');
      date = `${year}-${month}-${day}`;
    }

    const url = `${this.nhlApiBase}/schedule/${date}`;

    try {
      console.log(`📡 Fetching NHL games for ${date}...`);
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const data = await response.json();
      const gameWeek = data.gameWeek || [];

      if (gameWeek.length === 0) {
        console.log(`❌ No games found for ${date}`);
        return [];
      }

      const games = gameWeek[0]?.games || [];
      console.log(`✅ Found ${games.length} games\n`);

      return await this.processGames(games, date);

    } catch (error) {
      console.error(`❌ Error fetching NHL data: ${error.message}`);
      return [];
    }
  }

  /**
   * Process raw NHL API game data into clean format
   */
  async processGames(games, date) {
    const processed = [];

    for (const game of games) {
      try {
        const gameData = {
          date: date,
          gameId: game.id,
          awayTeam: game.awayTeam.abbrev,
          homeTeam: game.homeTeam.abbrev,
          awayScore: game.awayTeam.score || 0,
          homeScore: game.homeTeam.score || 0,
          totalScore: (game.awayTeam.score || 0) + (game.homeTeam.score || 0),
          gameState: game.gameState || 'FUT',  // FUT, LIVE, FINAL, OFF
          period: game.periodDescriptor?.number || 0,
          periodDescriptor: game.periodDescriptor?.periodType || '',
          clock: '',
          gameTime: game.startTimeUTC || '',
          lastUpdate: new Date().toISOString()
        };

        // Determine status
        // CRIT = Critical (final moments/just ended), OFF = Officially Over, FINAL = Final
        if (['OFF', 'FINAL', 'CRIT'].includes(gameData.gameState)) {
          gameData.status = 'FINAL';
        } else if (gameData.gameState === 'LIVE') {
          gameData.status = 'LIVE';
          // Fetch detailed clock data for LIVE games
          await this.fetchLiveGameDetails(gameData);
        } else if (['FUT', 'PRE'].includes(gameData.gameState)) {
          gameData.status = 'SCHEDULED';
        } else {
          gameData.status = 'UNKNOWN';
        }

        processed.push(gameData);

        // Print game info
        const statusEmoji = {
          'LIVE': '🔴',
          'FINAL': '✅',
          'SCHEDULED': '📅',
          'UNKNOWN': '❓'
        };

        const emoji = statusEmoji[gameData.status] || '❓';
        const periodInfo = gameData.status === 'LIVE' && gameData.period > 0
          ? ` P${gameData.period} ${gameData.clock}`
          : '';
        
        console.log(`${emoji} ${gameData.awayTeam} @ ${gameData.homeTeam}: ${gameData.awayScore}-${gameData.homeScore} (${gameData.status})${periodInfo}`);

      } catch (error) {
        console.warn(`⚠️  Skipping malformed game data: ${error.message}`);
        continue;
      }
    }

    return processed;
  }

  /**
   * Save game data to JSON file
   */
  async saveToJson(games) {
    try {
      const data = {
        lastUpdate: new Date().toISOString(),
        games: games,
        gamesCount: games.length
      };

      await fs.writeFile(
        this.outputFile,
        JSON.stringify(data, null, 2),
        'utf-8'
      );

      console.log(`✅ Saved ${games.length} games to ${this.outputFile}`);
      return true;

    } catch (error) {
      console.error(`❌ Error saving to JSON: ${error.message}`);
      return false;
    }
  }

  /**
   * Save game data to Firebase Firestore
   */
  async saveToFirebase(games) {
    if (!this.db) {
      console.log('⏭️  Skipping Firebase (not initialized)');
      return true;
    }

    try {
      const data = {
        lastUpdate: new Date().toISOString(),
        games: games,
        gamesCount: games.length
      };

      await this.db.collection('live_scores').doc('current').set(data);

      console.log(`✅ Saved ${games.length} games to Firebase`);
      return true;

    } catch (error) {
      console.error(`❌ Error saving to Firebase: ${error.message}`);
      return false;
    }
  }

  /**
   * Main update method
   */
  async update(date = null) {
    const games = await this.fetchGames(date);

    if (games.length === 0) {
      return false;
    }

    // Save to both JSON and Firebase
    const jsonSuccess = await this.saveToJson(games);
    const firebaseSuccess = await this.saveToFirebase(games);

    return jsonSuccess && firebaseSuccess;
  }
}

/**
 * Parse command line arguments
 */
function parseArgs() {
  const args = process.argv.slice(2);
  const config = {
    date: null,
    continuous: false,
    interval: 5 // minutes
  };

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--date' && args[i + 1]) {
      config.date = args[i + 1];
      i++;
    } else if (args[i] === '--continuous') {
      config.continuous = true;
    } else if (args[i] === '--interval' && args[i + 1]) {
      config.interval = parseInt(args[i + 1]);
      i++;
    } else if (args[i] === '--help' || args[i] === '-h') {
      console.log(`
NHL Live Score Updater

Usage:
  node updateLiveScores.js [options]

Options:
  --date YYYY-MM-DD    Fetch games for specific date (default: today)
  --continuous         Run continuously every N minutes
  --interval N         Minutes between updates (default: 5)
  --help, -h           Show this help message

Examples:
  node updateLiveScores.js
  node updateLiveScores.js --date 2025-10-23
  node updateLiveScores.js --continuous
  node updateLiveScores.js --continuous --interval 10
`);
      process.exit(0);
    }
  }

  return config;
}

/**
 * Sleep for specified milliseconds
 */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Main execution
 */
async function main() {
  const config = parseArgs();
  const updater = new NHLScoreUpdater();

  console.log('🏒 NHL Live Score Updater');
  console.log('='.repeat(50));
  console.log();

  if (config.continuous) {
    console.log(`🔄 Running continuously (every ${config.interval} minutes)`);
    console.log('   Press Ctrl+C to stop\n');

    try {
      while (true) {
        await updater.update(config.date);

        const now = new Date();
        const nextUpdate = new Date(now.getTime() + config.interval * 60000);
        
        console.log(`\n⏰ Next update at ${nextUpdate.toLocaleTimeString()}`);
        console.log(`   Waiting ${config.interval} minutes...\n`);
        
        await sleep(config.interval * 60000);
      }
    } catch (error) {
      if (error.message.includes('aborted')) {
        console.log('\n\n👋 Stopped by user');
        process.exit(0);
      }
      throw error;
    }
  } else {
    // Single update
    const success = await updater.update(config.date);
    process.exit(success ? 0 : 1);
  }
}

// Handle Ctrl+C gracefully
process.on('SIGINT', () => {
  console.log('\n\n👋 Stopped by user');
  process.exit(0);
});

// Run
main().catch(error => {
  console.error('❌ Fatal error:', error);
  process.exit(1);
});

File: update_live_scores.py
#!/usr/bin/env python3
"""
NHL Live Score Updater
======================
Fetches live scores from NHL API and updates a JSON file for display.
Does NOT rescrape odds - only updates scores 2-3 times during game time.

Usage:
  python update_live_scores.py                 # Update today's games
  python update_live_scores.py --date 2025-10-23  # Specific date
  python update_live_scores.py --continuous    # Run every 5 minutes

Requirements:
  pip install requests python-dotenv firebase-admin
"""

import requests
import json
import os
import sys
import time
import argparse
from datetime import datetime, timedelta
from pathlib import Path

# Optional Firebase support (if you want to also update Firebase)
FIREBASE_ENABLED = False
try:
    import firebase_admin
    from firebase_admin import credentials, firestore
    from dotenv import load_dotenv
    FIREBASE_ENABLED = True
except ImportError:
    print("⚠️  Firebase not installed. Score updates will only go to JSON file.")
    print("   To enable Firebase: pip install firebase-admin python-dotenv\n")


class NHLScoreUpdater:
    """Fetches live NHL scores and updates local JSON file"""
    
    def __init__(self, output_dir='public', firebase_enabled=False):
        self.output_dir = Path(output_dir)
        self.output_file = self.output_dir / 'live_scores.json'
        self.nhl_api_base = 'https://api-web.nhle.com/v1'
        self.firebase_enabled = firebase_enabled and FIREBASE_ENABLED
        self.db = None
        
        # Ensure output directory exists
        self.output_dir.mkdir(exist_ok=True)
        
        # Initialize Firebase if requested
        if self.firebase_enabled:
            self._init_firebase()
    
    def _init_firebase(self):
        """Initialize Firebase Admin SDK"""
        try:
            load_dotenv()
            
            # Check if already initialized
            if not firebase_admin._apps:
                cred = credentials.Certificate({
                    "type": "service_account",
                    "project_id": os.getenv('VITE_FIREBASE_PROJECT_ID'),
                    "private_key": os.getenv('FIREBASE_PRIVATE_KEY', '').replace('\\n', '\n'),
                    "client_email": os.getenv('FIREBASE_CLIENT_EMAIL'),
                })
                firebase_admin.initialize_app(cred)
            
            self.db = firestore.client()
            print("✅ Firebase initialized")
        except Exception as e:
            print(f"⚠️  Firebase initialization failed: {e}")
            self.firebase_enabled = False
    
    def fetch_games(self, date=None):
        """
        Fetch game data from NHL API for specified date
        
        Args:
            date (str): Date in YYYY-MM-DD format. Defaults to today.
        
        Returns:
            list: List of game dictionaries
        """
        if date is None:
            date = datetime.now().strftime('%Y-%m-%d')
        
        url = f'{self.nhl_api_base}/schedule/{date}'
        
        try:
            print(f"📡 Fetching NHL games for {date}...")
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            game_week = data.get('gameWeek', [])
            
            if not game_week:
                print(f"❌ No games found for {date}")
                return []
            
            games = game_week[0].get('games', [])
            print(f"✅ Found {len(games)} games\n")
            
            return self._process_games(games, date)
            
        except requests.RequestException as e:
            print(f"❌ Error fetching NHL data: {e}")
            return []
    
    def _process_games(self, games, date):
        """Process raw NHL API game data into clean format"""
        processed = []
        
        for game in games:
            try:
                game_data = {
                    'date': date,
                    'gameId': game.get('id'),
                    'awayTeam': game['awayTeam']['abbrev'],
                    'homeTeam': game['homeTeam']['abbrev'],
                    'awayScore': game['awayTeam'].get('score', 0),
                    'homeScore': game['homeTeam'].get('score', 0),
                    'totalScore': game['awayTeam'].get('score', 0) + game['homeTeam'].get('score', 0),
                    'gameState': game.get('gameState', 'FUT'),  # FUT, LIVE, FINAL, OFF
                    'period': game.get('period', 0),
                    'periodDescriptor': game.get('periodDescriptor', {}).get('periodType', ''),
                    'clock': game.get('clock', {}).get('timeRemaining', ''),
                    'gameTime': game.get('startTimeUTC', ''),
                    'lastUpdate': datetime.now().isoformat()
                }
                
                # Determine status
                if game_data['gameState'] in ['OFF', 'FINAL']:
                    game_data['status'] = 'FINAL'
                elif game_data['gameState'] == 'LIVE':
                    game_data['status'] = 'LIVE'
                elif game_data['gameState'] in ['FUT', 'PRE']:
                    game_data['status'] = 'SCHEDULED'
                else:
                    game_data['status'] = 'UNKNOWN'
                
                processed.append(game_data)
                
                # Print game info
                status_emoji = {
                    'LIVE': '🔴',
                    'FINAL': '✅',
                    'SCHEDULED': '📅',
                    'UNKNOWN': '❓'
                }
                
                emoji = status_emoji.get(game_data['status'], '❓')
                print(f"{emoji} {game_data['awayTeam']} @ {game_data['homeTeam']}: "
                      f"{game_data['awayScore']}-{game_data['homeScore']} ({game_data['status']})")
                
            except KeyError as e:
                print(f"⚠️  Skipping malformed game data: missing {e}")
                continue
        
        return processed
    
    def save_to_json(self, games):
        """Save game data to JSON file"""
        try:
            data = {
                'lastUpdate': datetime.now().isoformat(),
                'games': games,
                'gamesCount': len(games)
            }
            
            with open(self.output_file, 'w') as f:
                json.dump(data, f, indent=2)
            
            print(f"\n✅ Saved {len(games)} games to {self.output_file}")
            return True
            
        except Exception as e:
            print(f"❌ Error saving to JSON: {e}")
            return False
    
    def update_firebase_scores(self, games):
        """Update scores in Firebase (optional)"""
        if not self.firebase_enabled or not self.db:
            return False
        
        try:
            print("\n📊 Updating Firebase...")
            updated = 0
            
            for game in games:
                # Only update games that have started or finished
                if game['status'] in ['LIVE', 'FINAL']:
                    # Find matching bets in Firebase
                    bets_ref = self.db.collection('bets')
                    query = bets_ref.where('game.awayTeam', '==', game['awayTeam']) \
                                   .where('game.homeTeam', '==', game['homeTeam']) \
                                   .where('status', '==', 'PENDING')
                    
                    docs = query.get()
                    
                    for doc in docs:
                        bet_data = doc.to_dict()
                        
                        # Calculate outcome if game is final
                        if game['status'] == 'FINAL':
                            outcome = self._calculate_outcome(game, bet_data['bet'])
                            profit = self._calculate_profit(outcome, bet_data['bet']['odds'])
                            
                            doc.reference.update({
                                'result.awayScore': game['awayScore'],
                                'result.homeScore': game['homeScore'],
                                'result.totalScore': game['totalScore'],
                                'result.winner': 'AWAY' if game['awayScore'] > game['homeScore'] else 'HOME',
                                'result.outcome': outcome,
                                'result.profit': profit,
                                'result.fetched': True,
                                'result.fetchedAt': firestore.SERVER_TIMESTAMP,
                                'result.source': 'NHL_API_LIVE',
                                'status': 'COMPLETED'
                            })
                            
                            print(f"   ✅ Updated {game['awayTeam']} @ {game['homeTeam']} → {outcome}")
                            updated += 1
                        else:
                            # Just update scores for live games
                            doc.reference.update({
                                'result.awayScore': game['awayScore'],
                                'result.homeScore': game['homeScore'],
                                'result.totalScore': game['totalScore'],
                                'result.fetchedAt': firestore.SERVER_TIMESTAMP,
                                'result.source': 'NHL_API_LIVE'
                            })
            
            print(f"✅ Updated {updated} bets in Firebase")
            return True
            
        except Exception as e:
            print(f"❌ Error updating Firebase: {e}")
            return False
    
    def _calculate_outcome(self, game, bet):
        """Calculate bet outcome"""
        total_score = game['totalScore']
        
        if bet['market'] == 'TOTAL':
            if bet['side'] == 'OVER':
                return 'WIN' if total_score > bet['line'] else 'LOSS' if total_score < bet['line'] else 'PUSH'
            else:
                return 'WIN' if total_score < bet['line'] else 'LOSS' if total_score > bet['line'] else 'PUSH'
        
        elif bet['market'] == 'MONEYLINE':
            winner = 'AWAY' if game['awayScore'] > game['homeScore'] else 'HOME'
            return 'WIN' if bet['side'] == winner else 'LOSS'
        
        elif bet['market'] in ['PUCK_LINE', 'PUCKLINE']:
            spread = bet.get('line', 1.5)
            if bet['side'] == 'HOME':
                home_spread = game['homeScore'] - game['awayScore']
                return 'WIN' if home_spread > spread else 'LOSS' if home_spread < spread else 'PUSH'
            else:
                away_spread = game['awayScore'] - game['homeScore']
                return 'WIN' if away_spread > spread else 'LOSS' if away_spread < spread else 'PUSH'
        
        return None
    
    def _calculate_profit(self, outcome, odds):
        """Calculate profit in units"""
        if outcome == 'PUSH':
            return 0
        if outcome == 'LOSS':
            return -1
        
        # WIN
        if odds < 0:
            return 100 / abs(odds)
        else:
            return odds / 100
    
    def update(self, date=None, update_firebase=False):
        """
        Main update method
        
        Args:
            date (str): Date to update (YYYY-MM-DD)
            update_firebase (bool): Whether to also update Firebase
        """
        games = self.fetch_games(date)
        
        if not games:
            return False
        
        # Always save to JSON
        json_success = self.save_to_json(games)
        
        # Optionally update Firebase
        if update_firebase and self.firebase_enabled:
            self.update_firebase_scores(games)
        
        return json_success


def main():
    parser = argparse.ArgumentParser(description='Update NHL live scores')
    parser.add_argument('--date', type=str, help='Date in YYYY-MM-DD format (default: today)')
    parser.add_argument('--continuous', action='store_true', help='Run continuously every 5 minutes')
    parser.add_argument('--firebase', action='store_true', help='Also update Firebase')
    parser.add_argument('--interval', type=int, default=5, help='Update interval in minutes (default: 5)')
    
    args = parser.parse_args()
    
    # Initialize updater
    updater = NHLScoreUpdater(firebase_enabled=args.firebase)
    
    print("🏒 NHL Live Score Updater")
    print("=" * 50)
    
    if args.continuous:
        print(f"🔄 Running continuously (every {args.interval} minutes)")
        print("   Press Ctrl+C to stop\n")
        
        try:
            while True:
                updater.update(args.date, update_firebase=args.firebase)
                print(f"\n⏰ Next update in {args.interval} minutes...")
                print(f"   {datetime.now().strftime('%I:%M:%S %p')}\n")
                time.sleep(args.interval * 60)
        except KeyboardInterrupt:
            print("\n\n👋 Stopped by user")
            sys.exit(0)
    else:
        # Single update
        success = updater.update(args.date, update_firebase=args.firebase)
        sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

